File: v2/libs/go/broker/broker.go
// Package broker membungkus NATS JetStream untuk publish/subscribe event.
package broker

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/nats-io/nats.go"

	"github.com/simpulx/v2/libs/go/events"
)

type Broker struct {
	nc *nats.Conn
	js nats.JetStreamContext
}

// Connect membuka koneksi NATS dan memastikan stream EVENTS ada.
func Connect(url string) (*Broker, error) {
	nc, err := nats.Connect(url,
		nats.RetryOnFailedConnect(true),
		nats.MaxReconnects(-1),
		nats.ReconnectWait(time.Second),
	)
	if err != nil {
		return nil, fmt.Errorf("nats connect: %w", err)
	}
	js, err := nc.JetStream()
	if err != nil {
		return nil, fmt.Errorf("jetstream: %w", err)
	}
	// Idempotent: buat stream bila belum ada.
	_, err = js.AddStream(&nats.StreamConfig{
		Name:      events.StreamName,
		Subjects:  []string{events.StreamSubjects},
		Retention: nats.LimitsPolicy,
		MaxAge:    7 * 24 * time.Hour,
		Storage:   nats.FileStorage,
	})
	if err != nil && err != nats.ErrStreamNameAlreadyInUse {
		// AddStream pada stream eksisting dgn config sama bisa mengembalikan error;
		// abaikan bila stream sudah ada.
		if _, infoErr := js.StreamInfo(events.StreamName); infoErr != nil {
			return nil, fmt.Errorf("add stream: %w", err)
		}
	}
	return &Broker{nc: nc, js: js}, nil
}

// Publish mengirim event ber-amplop. Bila orgID/data sudah disediakan caller,
// fungsi mengisi id & ts otomatis.
func (b *Broker) Publish(subject, orgID string, payload any) error {
	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	env := events.Envelope{
		ID:    uuid.NewString(),
		Type:  subjectToType(subject),
		OrgID: orgID,
		TS:    time.Now().UTC(),
		Data:  data,
	}
	raw, err := json.Marshal(env)
	if err != nil {
		return err
	}
	_, err = b.js.Publish(subject, raw)
	return err
}

// Subscribe membuat durable queue consumer. handler menerima envelope yang
// sudah ter-decode; kembalikan error untuk Nak (redelivery).
func (b *Broker) Subscribe(subject, durable string, handler func(events.Envelope) error) error {
	_, err := b.js.QueueSubscribe(subject, durable, func(m *nats.Msg) {
		var env events.Envelope
		if err := json.Unmarshal(m.Data, &env); err != nil {
			_ = m.Term() // payload rusak, jangan redeliver
			return
		}
		if err := handler(env); err != nil {
			_ = m.Nak()
			return
		}
		_ = m.Ack()
	},
		nats.Durable(durable),
		nats.ManualAck(),
		nats.AckWait(30*time.Second),
		nats.DeliverAll(),
	)
	return err
}

func (b *Broker) Close() { b.nc.Close() }

func subjectToType(subject string) string {
	// "events.message.received" -> "message.received"
	if len(subject) > len("events.") {
		return subject[len("events."):]
	}
	return subject
}

File: v2/libs/python/simpulx_common/broker.py
"""Wrapper NATS JetStream untuk service Python (kompatibel dgn amplop Go)."""
from __future__ import annotations

import json
import uuid
from datetime import datetime, timezone
from typing import Awaitable, Callable

import nats
from nats.js import JetStreamContext
from nats.js.api import ConsumerConfig


class Broker:
    def __init__(self, nc, js: JetStreamContext):
        self.nc = nc
        self.js = js

    @classmethod
    async def connect(cls, url: str) -> "Broker":
        nc = await nats.connect(url, reconnect_time_wait=1, max_reconnect_attempts=-1)
        js = nc.jetstream()
        return cls(nc, js)

    async def publish(self, subject: str, org_id: str, data: dict) -> None:
        envelope = {
            "id": str(uuid.uuid4()),
            "type": subject.removeprefix("events."),
            "org_id": org_id,
            "ts": datetime.now(timezone.utc).isoformat(),
            "data": data,
        }
        await self.js.publish(subject, json.dumps(envelope).encode())

    async def subscribe(
        self, subject: str, durable: str, handler: Callable[[dict], Awaitable[bool]]
    ) -> None:
        """Push subscription durable+queue. handler menerima ENVELOPE (dict).
        Kembalikan True untuk ack, False untuk nak (redeliver)."""

        async def cb(msg):
            try:
                env = json.loads(msg.data.decode())
            except Exception:  # noqa: BLE001
                await msg.term()  # payload rusak
                return
            try:
                ok = await handler(env)
            except Exception:  # noqa: BLE001
                await msg.nak()
                return
            if ok:
                await msg.ack()
            else:
                await msg.nak()

        config = ConsumerConfig(
            ack_wait=30.0,
            max_deliver=3,
        )
        await self.js.subscribe(
            subject, durable=durable, queue=durable, cb=cb, manual_ack=True, config=config
        )

    async def close(self) -> None:
        await self.nc.drain()
